Update release notes with Ballerina Shell changes (#105)
ballerina-platform/ballerina-release: release-notes/swan-lake-alpha3-release-note.md

##### Language Server

##### Ballerina Shell

> - Ballerina Shell now supports redefining module-level definitions and variable declarations. 

Reviewer: Shall we add the example here?
````suggestion
- Ballerina Shell now supports redefining module-level definitions and variable declarations. 
```ballerina
=$ int i = 3;
=$ string i = "Hello";  // Same variable can be redefined
=$ i
"Hello"
```
````

- The `/remove` command can be used to remove one or more declarations from snippet memory.

```ballerina
=$ int i = 3;
=$ string j = "Hi";

> =$ string i = "Hello";  // Same variable can be redefined

Reviewer: If we split the example

```suggestion
```

=$ /remove i j   // Variables can be removed
=$ i
| error: undefined symbol 'i'
|       i
|       ^
| Compilation aborted due to errors.
```

- Ballerina Shell can now load definitions and declarations from a file. The file to load from can be specified using the `-f` or `--file` command-line options when launching Ballerina Shell. Alternatively the `/file` command can also be used for this purpose, from within the shell. 

```bash
$ bal shell -f my_file.bal
```

> - The `--force-dumb` command-line option will now have only a long option and the short option `-f` is now used to load from a file.

Reviewer: This doesn't have to be a separate bullet point, right?
```suggestion
The `--force-dumb` command-line option will now have only a long option and the short option `-f` is now used to load from a file.
```

> - Ballerina Shell now supports qualifiers, cyclic type definitions, and list binding patterns.

Reviewer: ```suggestion
- Ballerina Shell now supports cyclic type definitions and list binding patterns.
```

##### Language Server

##### Ballerina Shell

> - Ballerina Shell now supports redefining module-level definitions and variable declarations. 

Reviewer: Isn't this supported with local variables?

Reviewer: Ah, doesn't make sense for local variables. This is OK.

> - Ballerina Shell now supports redefining module-level definitions and variable declarations. 

Reviewer 2: ```suggestion
- The Ballerina Shell now supports redefining module-level definitions and variable declarations. 
```

```ballerina
=$ int i = 3;
=$ string j = "Hi";
=$ string i = "Hello";  // Same variable can be redefined
```

> - The `/remove` command can be used to remove one or more declarations from snippet memory.

Reviewer: ```suggestion
- A new `/remove` command has been introduced to be used from within the Ballerina Shell to remove one or more declarations from snippet memory.
```

> - A new `/remove` command has been introduced to be used from within the Ballerina Shell to remove one or more declarations from snippet memory.

Reviewer 2: ```suggestion
- A new `/remove` command has been introduced to be used from within the Ballerina Shell to remove one or more declarations from the snippet memory.
```

```ballerina
=$ int i = 3;
=$ string j = "Hi";
=$ /remove i j
=$ i
| error: undefined symbol 'i'
|       i
|       ^
| Compilation aborted due to errors.
```

> - Ballerina Shell can now load definitions and declarations from a file. The file to load from can be specified using the `-f` or `--file` command-line options when launching Ballerina Shell. Alternatively the `/file` command can also be used for this purpose, from within the shell. 

Reviewer 2: ```suggestion
- Ballerina Shell can now load definitions and declarations from a file. The file to load from can be specified using the `-f` or `--file` command-line options when launching the Ballerina Shell. Alternatively, the `/file` command can also be used for this purpose from within the Shell. 
```

```bash
$ bal shell -f my_file.bal
```

The `--force-dumb` command-line option will now have only a long option and the short option `-f` is now used to load from a file.

> - Ballerina Shell now supports cyclic type definitions and list binding patterns.

Reviewer 2: ```suggestion
- The Ballerina Shell now supports cyclic-type definitions and list binding patterns.
```

> - The Ballerina Shell now supports cyclic-type definitions and list binding patterns.

Reviewer: Do we need the hyphen here? IMO, it is the definition itself that is cyclic, not the type.

@praneesha @kdsuneraavinash

Author: yeah, IMO hyphen is not required

Reviewer 2: +1 to not have the hyphen.

> - Ballerina Shell now preserves qualifiers such as the `final` qualifier of a variable declaration.

Reviewer 2: ```suggestion
- The Ballerina Shell now preserves qualifiers such as the `final` qualifier of a variable declaration.
```

##### Language Server

##### Ballerina Shell

> Ballerina Shell now supports redefining module-level declarations and variable declarations. `/remove NAMES` command can be used to remove one or multiple declarations from snippet memory.

Reviewer: By module-level declarations I'm assuming you mean functions, types, etc.? Shall we say definitions instead?
```suggestion
- Ballerina Shell now supports redefining module-level definitions and variable declarations. The `/remove` command can be used to remove one or more declarations from snippet memory.
```

Also need to use bullet points, right?

> Ballerina Shell now supports redefining module-level declarations and variable declarations. `/remove NAMES` command can be used to remove one or multiple declarations from snippet memory.

Reviewer: Shall we split this into two? First mention redefining then add the snippet and then add the content on remove and the snippet?

> ```ballerina

Reviewer: Should we use bash here too?

Author: Im okay with both bash and ballerina. I used Ballerina because most of the code in the snippet is ballerina code.

=$ int i = 3;
=$ string j = "Hi";
=$ string i = "Hello";  // Same variable can be redefined
=$ /remove i j   // Variables can be removed
=$ i
| error: undefined symbol 'i'
|       i
|       ^
| Compilation aborted due to errors.
```

> A command-line argument was added with `-f` or `--file` argument name to load module-level declarations within a file. `/file <FILENAME` command can also be used for this purpose, from within the shell. Note that `--force-debug` will now have only a long option; `-f` short option is now used to open the file.

Reviewer 3: ```suggestion
Module-level declarations within a file can be loaded using the `-f` or `--file` command-line arguments. `/file <FILENAME` command can also be used for this purpose, from within the shell. Note that `--force-debug` will now have only a long option; `-f` short option is now used to open the file.

Reviewer 3: Maryam's suggestion sounds better. Please ignore this.

> A command-line argument was added with `-f` or `--file` argument name to load module-level declarations within a file. `/file <FILENAME` command can also be used for this purpose, from within the shell. Note that `--force-debug` will now have only a long option; `-f` short option is now used to open the file.

Reviewer: Shall we emphasize the feature first?

```suggestion
- Ballerina Shell can now load definitions and declarations from a file. The file to load from can be specified using the `-f` or `--file` command-line options when launching Ballerina Shell. Alternatively the `/file` command can also be used for this purpose, from within the shell. 

Note that the `--force-dumb` command-line option will now have only a long option and the short option `-f` is now used to load from a file.
```

Should be `--force-dumb` right? Not `--force-debug`?

```bash
$ bal shell -f my_file.bal
```

Also, error messages for error objects, panics, and fails will show distinct error messages from each other. Panics will show `panic: ERROR` and fails will show `fail ERROR`. Error-values will simply be evaluated to their value.

```ballerina
=$ error("Error") // Error values will simply evaluate
error("Error")
=$ panic error("Error")  // Panics will show as panic runtime errors
panic: {ballerina}DivisionByZero {"message":" / by zero"}
| Execution aborted due to unhandled runtime error.
=$ fail error("Error")  // Fails will show as fails
fail: Error {}

> ```

Reviewer: Do we need this in the release note? It's not a feature, right?

> Qualifiers (e.g., `final`) will now correctly work and `public` will declarations are allowed. Additionally, Ballerina Shell will now correctly work with cyclic type definitions and array binding patterns.

Reviewer 3: `Qualifiers (e.g., `final`) will now correctly work`
Can you explain how it's correctly working?

> Qualifiers (e.g., `final`) will now correctly work and `public` will declarations are allowed. Additionally, Ballerina Shell will now correctly work with cyclic type definitions and array binding patterns.

Reviewer: ```suggestion
- Ballerina Shell now supports qualifiers, cyclic type definitions, and list binding patterns.
```

Reviewer: Or
```suggestion
- Ballerina Shell now supports cyclic type definitions and list binding patterns.

- Ballerina Shell now preserves qualifiers such as the `final` qualifier of a variable declaration.
```

> A bug causing some snippets (e.g., `json x = {`) to be incorrectly identified as complete, was fixed.

Reviewer: We don't have to mention bug fixes unless they are breaking changes.

> A bug causing some snippets (e.g., `json x = {`) to be incorrectly identified as complete, was fixed.

Reviewer 3: Maybe we can word this differently rather than saying a bug was fixed? Maybe include how it's identified now?
